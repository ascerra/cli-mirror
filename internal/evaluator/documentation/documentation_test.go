// Copyright The Enterprise Contract Contributors
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
// SPDX-License-Identifier: Apache-2.0

//go:build unit

package main

import (
	"io/fs"
	"os"
	"path/filepath"
	"testing"

	"github.com/gkampitakis/go-snaps/snaps"
	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"
)

func TestWriteBuiltinsToYAML(t *testing.T) {
	dir := t.TempDir()
	err := writeBultinsToYAML(dir)
	require.NoError(t, err)
	found := 0
	err = filepath.WalkDir(dir, func(path string, d fs.DirEntry, err error) error {
		require.NoError(t, err)
		if d.IsDir() {
			return nil
		}
		contents, err := os.ReadFile(path)
		found += 1
		require.NoError(t, err)
		snaps.MatchSnapshot(t, string(contents))
		return nil
	})
	require.NoError(t, err)
	assert.NotZero(t, found, "did not find any ec builtins")
}
