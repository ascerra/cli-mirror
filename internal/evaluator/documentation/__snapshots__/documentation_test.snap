
[TestWriteBuiltinsToYAML - 1]
decl:
  args:
  - description: OCI blob reference
    name: ref
    type: string
  result:
    description: the OCI blob
    name: blob
    type: string
  type: function
description: Fetch a blob from an OCI registry.
name: ec.oci.blob
nondeterministic: true

---

[TestWriteBuiltinsToYAML - 2]
decl:
  args:
  - description: OCI image reference
    name: ref
    type: string
  result:
    description: the Image Manifest object
    name: object
    static:
    - key: annotations
      value:
        dynamic:
          key:
            type: string
          value:
            type: string
        type: object
    - key: config
      value:
        static:
        - key: annotations
          value:
            dynamic:
              key:
                type: string
              value:
                type: string
            type: object
        - key: artifactType
          value:
            type: string
        - key: data
          value:
            type: string
        - key: digest
          value:
            type: string
        - key: mediaType
          value:
            type: string
        - key: platform
          value:
            static:
            - key: architecture
              value:
                type: string
            - key: features
              value:
                static:
                - type: string
                type: array
            - key: os
              value:
                type: string
            - key: os.features
              value:
                static:
                - type: string
                type: array
            - key: os.version
              value:
                type: string
            - key: variant
              value:
                type: string
            type: object
        - key: size
          value:
            type: number
        - key: urls
          value:
            static:
            - type: string
            type: array
        type: object
    - key: layers
      value:
        static:
        - static:
          - key: annotations
            value:
              dynamic:
                key:
                  type: string
                value:
                  type: string
              type: object
          - key: artifactType
            value:
              type: string
          - key: data
            value:
              type: string
          - key: digest
            value:
              type: string
          - key: mediaType
            value:
              type: string
          - key: platform
            value:
              static:
              - key: architecture
                value:
                  type: string
              - key: features
                value:
                  static:
                  - type: string
                  type: array
              - key: os
                value:
                  type: string
              - key: os.features
                value:
                  static:
                  - type: string
                  type: array
              - key: os.version
                value:
                  type: string
              - key: variant
                value:
                  type: string
              type: object
          - key: size
            value:
              type: number
          - key: urls
            value:
              static:
              - type: string
              type: array
          type: object
        type: array
    - key: mediaType
      value:
        type: string
    - key: schemaVersion
      value:
        type: number
    - key: subject
      value:
        static:
        - key: annotations
          value:
            dynamic:
              key:
                type: string
              value:
                type: string
            type: object
        - key: artifactType
          value:
            type: string
        - key: data
          value:
            type: string
        - key: digest
          value:
            type: string
        - key: mediaType
          value:
            type: string
        - key: platform
          value:
            static:
            - key: architecture
              value:
                type: string
            - key: features
              value:
                static:
                - type: string
                type: array
            - key: os
              value:
                type: string
            - key: os.features
              value:
                static:
                - type: string
                type: array
            - key: os.version
              value:
                type: string
            - key: variant
              value:
                type: string
            type: object
        - key: size
          value:
            type: number
        - key: urls
          value:
            static:
            - type: string
            type: array
        type: object
    type: object
  type: function
description: Fetch an Image Manifest from an OCI registry.
name: ec.oci.image_manifest
nondeterministic: true

---

[TestWriteBuiltinsToYAML - 3]
decl:
  args:
  - description: the PURL
    name: purl
    type: string
  result:
    description: PURL validity
    name: result
    type: string
  type: function
description: Determine whether or not a given PURL is valid.
name: ec.purl.is_valid

---

[TestWriteBuiltinsToYAML - 4]
decl:
  args:
  - description: the PURL
    name: purl
    type: string
  result:
    description: the parsed PURL object
    name: object
    static:
    - key: name
      value:
        type: string
    - key: namespace
      value:
        type: string
    - key: qualifiers
      value:
        dynamic:
          static:
          - key: key
            value:
              type: string
          - key: value
            value:
              type: string
          type: object
        type: array
    - key: subpath
      value:
        type: string
    - key: type
      value:
        type: string
    - key: version
      value:
        type: string
    type: object
  type: function
description: Parse a valid PURL into an object.
name: ec.purl.parse

---

[TestWriteBuiltinsToYAML - 5]
decl:
  args:
  - description: OCI image reference
    name: ref
    type: string
  - description: Sigstore verification options
    name: opts
    static:
    - key: certificate_identity
      value:
        type: string
    - key: certificate_identity_regexp
      value:
        type: string
    - key: certificate_oidc_issuer
      value:
        type: string
    - key: certificate_oidc_issuer_regexp
      value:
        type: string
    - key: ignore_rekor
      value:
        type: boolean
    - key: public_key
      value:
        type: string
    - key: rekor_url
      value:
        type: string
    type: object
  result:
    description: the result of the verification request
    name: result
    static:
    - key: attestations
      value:
        description: matching attestations
        name: attestations
        static:
        - description: attestation matching provided identity/key
          name: attestation
          static:
          - key: signatures
            value:
              description: signatures associated with attestation
              name: signatures
              static:
              - static:
                - key: certificate
                  value:
                    type: string
                - key: chain
                  value:
                    static:
                    - type: string
                    type: array
                - key: keyid
                  value:
                    type: string
                - key: metadata
                  value:
                    dynamic:
                      key:
                        type: string
                      value:
                        type: string
                    type: object
                - key: signature
                  value:
                    type: string
                type: object
              type: array
          - key: statement
            value:
              description: statement from attestation
              name: statement
              type: any
          type: object
        type: array
    - key: errors
      value:
        description: verification errors
        name: errors
        static:
        - type: string
        type: array
    - key: success
      value:
        description: true when verification is successful
        name: success
        type: boolean
    type: object
  type: function
description: Use sigstore to verify the attestation of an image.
name: ec.sigstore.verify_attestation
nondeterministic: true

---

[TestWriteBuiltinsToYAML - 6]
decl:
  args:
  - description: OCI image reference
    name: ref
    type: string
  - description: Sigstore verification options
    name: opts
    static:
    - key: certificate_identity
      value:
        type: string
    - key: certificate_identity_regexp
      value:
        type: string
    - key: certificate_oidc_issuer
      value:
        type: string
    - key: certificate_oidc_issuer_regexp
      value:
        type: string
    - key: ignore_rekor
      value:
        type: boolean
    - key: public_key
      value:
        type: string
    - key: rekor_url
      value:
        type: string
    type: object
  result:
    description: the result of the verification request
    name: result
    static:
    - key: errors
      value:
        description: verification errors
        name: errors
        static:
        - type: string
        type: array
    - key: signatures
      value:
        description: matching signatures
        name: signatures
        static:
        - static:
          - key: certificate
            value:
              type: string
          - key: chain
            value:
              static:
              - type: string
              type: array
          - key: keyid
            value:
              type: string
          - key: metadata
            value:
              dynamic:
                key:
                  type: string
                value:
                  type: string
              type: object
          - key: signature
            value:
              type: string
          type: object
        type: array
    - key: success
      value:
        description: true when verification is successful
        name: success
        type: boolean
    type: object
  type: function
description: Use sigstore to verify the signature of an image.
name: ec.sigstore.verify_image
nondeterministic: true

---
